# Exploring eBay Car Sales Data

i will work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was downloaded from Kaggle, which was originally scraped by a user.

## Summary of Findings

## Dataset Overview

The original dataset was cleaned to make the data easier to work with. However, our version has been intentionally dirtied, with a sample of 50,000 data points from the original (credits to DataQuest). If you need to refer to the original dataset, you can find it here.

## Data Dictionary
Here is an overview of the 20 columns in the dataset:

`dateCrawled`:- When this ad was first crawled. All field-values are taken from this date.

`name`:- Name of the car.

`seller`:- Whether the seller is private or a dealer.

`offerType`:- The type of listing

`price`:- The price on the ad to sell the car.

`abtest`:- Whether the listing is included in an A/B test.

`vehicleType`:- The vehicle Type.

`yearOfRegistration` - The year in which the car was first registered.

`gearbox`:- The transmission type.

`powerPS`:- The power of the car in PS.

`model`:- The car model name.

`kilometer`:- How many kilometers the car has driven.

`monthOfRegistration`:- The month in which the car was first registered.

`fuelType`:- What type of fuel the car uses.

`brand`:- The brand of the car.

`notRepairedDamage`:- If the car has a damage which is not yet repaired.

`dateCreated`:- The date on which the eBay listing was created.

`nrOfPictures`:- The number of pictures in the ad.

`postalCode`:- The postal code for the location of the vehicle.

`lastSeenOnline`:- When the crawler saw this ad last online.

Since our dataset version is dirty, we will perform some cleaning before analysis.


In [1]:
#import the dataset and display the first five rows

import pandas as pd
import numpy as np
from tabulate import tabulate
import matplotlib.pyplot as plt

autos = pd.read_csv('autos.csv', encoding="Latin-1")
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [2]:
#get more info on the dataset
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### Observations
* Our Dataset contains **20 columns** and **50,000 rows**, in which strings and integer datatypes are included.
* `price` and `odometer` columns contain numeric values stored as text.
* We have five columns with null values
* Columns with date informations such as `dateCrawled`,`dateCreated`, `lastSeen` are saved as objects instead of date.
* Rather than python's default snake case, the column names follow camelcase convention. 

As we explore further, we will try to correct these occurences, so they don't cause us problems later. 


## Data Cleaning
### 1. Cleaning column names

We will start by converting the column names from camelcase to snakecase. Let's print these columns to identify the ones we need to rename.


In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offertype', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'post_code',
       'last_seen']
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offertype   price  ab_test vehicle_type  registration_year   gear_box  \
0   Angebot  $5,000  control          bus               2004    manuell   
1   Angebot  $8,500  control    limousine               1997  automatik   
2   Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  num_pictures  post_code  \
0              nein  2016-03-26 00:00:00             0      79588   
1              nein  2016-04-04 00:00:00             0      71034   
2              nein  2016-03-26 00:00:00             0      35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

Let's explore further to determine what other cleaning tasks need to be done. We will try to identify:

Text columns where all or almost all values are the same. We can often drop these as they don't have helpful information for analysis.

Data that is not stored in the best format for analysis. They may need to be cleaned, converted or both.

We will use the `DataFrame.describe(include='all')` method to get information for categorical and numeric columns, then use the S`eries.value_counts()` and `Series.head()` methods to glean more insights if any columns need examining.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offertype  price ab_test  \
count                 50000        50000   50000     50000  50000   50000   
unique                48213        38754       2         2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0    test   
freq                      3           78   49999     49999   1421   25756   
mean                    NaN          NaN     NaN       NaN    NaN     NaN   
std                     NaN          NaN     NaN       NaN    NaN     NaN   
min                     NaN          NaN     NaN       NaN    NaN     NaN   
25%                     NaN          NaN     NaN       NaN    NaN     NaN   
50%                     NaN          NaN     NaN       NaN    NaN     NaN   
75%                     NaN          NaN     NaN       NaN    NaN     NaN   
max                     NaN          NaN     NaN       NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures     post_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

The columns `seller`, `offertype`, `ab_test`, `gear_box`, `unrepaired_damage` seems to have only two unique values.

In [6]:
#loop through each column and print the frequency of unique values

for item in ["seller", "offertype", "ab_test", "gear_box", "unrepaired_damage" ]:
    print('-'*30)
    print(autos[item].value_counts())
    print('-'*30)

------------------------------
privat        49999
gewerblich        1
Name: seller, dtype: int64
------------------------------
------------------------------
Angebot    49999
Gesuch         1
Name: offertype, dtype: int64
------------------------------
------------------------------
test       25756
control    24244
Name: ab_test, dtype: int64
------------------------------
------------------------------
manuell      36993
automatik    10327
Name: gear_box, dtype: int64
------------------------------
------------------------------
nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64
------------------------------


Observations
The Seller and offer_type columns are heavily one-sided:

In the seller column, Privat means 'private' and gewerblick means 'commercial'. This tells us that most records in the dataset were from private sellers.
In the offer_type column, Angebot stands for 'offer', while gesuch stands for 'request'. This tells us that majority of the records are car offered for sale.
These columns are far too one-sided to add helpful information to our analysis. We will drop them, but first, let's explore the `num_pictures` column.

In [7]:
autos['num_pictures'].value_counts()

0    50000
Name: num_pictures, dtype: int64

All data in the `num_pictures` column is zero. We will drop this column along with `seller`, `offertype`

In [8]:
print(autos.shape)
autos = autos.drop(['num_pictures','seller','offertype'],axis=1)
print(autos.shape)

(50000, 20)
(50000, 17)


## Translating Non-English words

In the previous observation, we manually interpreted the meaning of the terms in the seller and offer_type columns. Repeating this process for all affected columns will be cumbersome in the long run.

In total, there were six columns recorded in German. However, we had already dropped two of the affected columns. We will explore the remaining four in detail:

In [9]:
german_columns = ['vehicle_type', 'gear_box','fuel_type', 'unrepaired_damage']

#print unique german words from each column

for g_labels in german_columns:
    print(' ')
    print(autos[g_labels].unique())

 
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
 
['manuell' 'automatik' nan]
 
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
 
['nein' nan 'ja']


We can use this information to build a dictionary of translations for each German word. Then, use a function convert_language() to iterate through each of the affected columns, while mapping the right English words from our dictionary. The `series.map()` method will handle the translation/mapping process for us:



In [10]:
# translator function
def translate_column(df,column,value_dict):
    '''translates values in df column using value_dict as reference'''
    df[column] = df[column].map(value_dict)


# A dictionary of translations
translations ={# vehicle type terms
               'bus':'bus',
               'limousine':'limousine',
               'kleinwagen': 'small car',
               'kombi': 'combo',
               'coupe':'coupe',
               'suv':'suv',
               'cabrio': 'cabrio',
               'andere': 'other',
               'cabrio': 'convertible',
               # gear box terms
               'manuell': 'manual',
               'automatik': 'automatic',
               # fuel type terms
               'lpg':'lpg',
               'benzin': 'gasoline',
               'diesel':'diesel',
               'cng':'natural gas',
               'hybrid':'hybrid',
               'elektro':'electric',
               'andere':'other',
               # unrepaired damage terms
               'nein':'no',
               'ja':'yes'       
}

Let's translate each of these columns. We will also print out the value counts before and after translation to ensure that the process ran adequately, without compromising our data:



In [11]:
for item in german_columns:
    print('BEFORE TRANSLATING', '\n')
    
    print(autos[item].value_counts(dropna=False))
    
    print('\n', 'TRANSLATING WORDS...')
    print('-'*30)
    
    translate_column(autos, item, translations)
    print(autos[item].value_counts(dropna=False))
    
    print('-'*30)
    print('-'*30)

BEFORE TRANSLATING 

limousine     12859
kleinwagen    10822
kombi          9127
NaN            5095
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

 TRANSLATING WORDS...
------------------------------
limousine      12859
small car      10822
combo           9127
NaN             5095
bus             4093
convertible     3061
coupe           2537
suv             1986
other            420
Name: vehicle_type, dtype: int64
------------------------------
------------------------------
BEFORE TRANSLATING 

manuell      36993
automatik    10327
NaN           2680
Name: gear_box, dtype: int64

 TRANSLATING WORDS...
------------------------------
manual       36993
automatic    10327
NaN           2680
Name: gear_box, dtype: int64
------------------------------
------------------------------
BEFORE TRANSLATING 

benzin     30107
diesel     14567
NaN         4482
lpg          691
cng           75
hybrid       

### Observations
* The German words have been converted to English, and the value counts remained the same after translation. We can conclude that the translation was executed correctly without compromising our data.

## 3. Reformating numeric data

As observed previously the `price` and `odometer` data should have been entered in numeric data types instead of object type. For each of these columns we will remove the non-numeric characters, convert the value to numeric and then use `dataframe.rename()` method to assign descriptive column labels.


In [12]:
#removing non-numeric characters from price column
autos["price"] = (autos["price"].str.replace("$","")
                                .str.replace(",","")
                                .astype(int)
                 )
autos.rename({'price':"usd_price"},inplace=True,axis=1)
autos["usd_price"].head()

C:\Users\donjo\AppData\Local\Temp/ipykernel_14052/3124408520.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = (autos["price"].str.replace("$","")


0    5000
1    8500
2    8990
3    4350
4    1350
Name: usd_price, dtype: int32

In [13]:
#removing non-numeric characters from odometer column
autos["odometer"] = (autos["odometer"]
                                    .str.replace("km","")
                                    .str.replace(",","")
                                    .astype(int)
                    )
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

## 4. Detecting Numeric Outliers
Let's continue by exploring the data for outliers. Specifically, we are looking for data that doesn't look right, especially in our numeric columns (`odometer_km` and `usd_price`). We will analyze these columns using the five number summary (min, Q1, median, Q2, Max) then look for any values that look unrealistically high or low.

In [14]:
autos[['odometer_km','usd_price']].describe()

odometer_km     usd_price
count   50000.000000  5.000000e+04
mean   125732.700000  9.840044e+03
std     40042.211706  4.811044e+05
min      5000.000000  0.000000e+00
25%    125000.000000  1.100000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  1.000000e+08

### Observations:
* ***Odometer readings:***
    * The median and 75th percentile is same as the maximum value(150,000km). The 25th percentile is set around 125,000km.
    * This indicates that 150,000 is not an outlier as the 75% of the listed cars lies in the mileage between 125,000 and 150,000 kms.
    
* ***Price range***
    * The maximum price is 100,000,000USD. This varies significantly from the median and 75th percentile value which is 2950 USD and 7200 USD respectively. The standard deviation is also considerably high closer to 480,000 USD.
    * The minimum price is zero, which is unusual and could effect our analysis.
    
We can further explore `odometer_km` by sorted value counts:

In [15]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

***Additional Observations***
* The least mileage is observed to be 5000. 0 km is not in the list because this is a classifieds for used cars.
* Higher mileage cars are more on the list than lower mileage cars.
* Over 64% of the entries have readings of 150000. This could imply that there is an upper limit in the odometer readings for cars accepted in the list. 

## 5. Removing Price Outliers

As observed earlier, the 25th, 50th and 75th percentile of the prices are 1,100, 2,950 and 7,200USD. We will use a selection method to create a “data fence” from our 25th to 75th percentiles. Any data beyond this fence will be considered an outlier.

We already have enough information to calculate our interquartile range IQR. We will use the equation below to determine our upper and lower price fences:

    Interquartile Range (IQR) = 75th Percentile - 25th Percentile
    Lower fence = 25th Percentile - 1.5*(IQR)
    Upper fence = 75th Percentile + 1.5*(IQR)
    
Lets define all these boundaries in the code cell below:

In [16]:
price_25, price_75 = (1100, 7200)

iqr = price_75 - price_25
lower_fence = price_25 - (1.5*iqr)
upper_fence = price_75 + (1.5*iqr)

print('Interquartile Range: ', iqr)
print('Lower Fence: ', lower_fence)
print('Upper Fence: ', upper_fence)


Interquartile Range:  6100
Lower Fence:  -8050.0
Upper Fence:  16350.0


Fences may extend the data sometimes, so it is fine if our lower fence go negative. We should be mindful, however, that a minimum value of 0 can also effect our analysis. In this case, it is advisible to eliminate entries with zero prices too.

Before we proceed, let's take a look at how many listings are recorded with zero price.

In [17]:
autos['usd_price'].value_counts().sort_index().head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: usd_price, dtype: int64

We can also see that there are listings with price mentioned at $1, $2. Looking in detail at the table, we can judge that these entries are made for exchange or purposes other than sale, as the german word "Tauschen" is included in some entries, which translates to exchange. As these prices cannot be considered as a reasonable price for a vehicle, these entries are better be removed from our analysis. Therefore, a price of $200 will be set as the lower limit. 

In [18]:
autos = autos[autos['usd_price'].between(200, upper_fence)]
autos['usd_price'].describe()

count    43875.000000
mean      4173.076239
std       3837.353134
min        200.000000
25%       1200.000000
50%       2800.000000
75%       6000.000000
max      16350.000000
Name: usd_price, dtype: float64

### Observations:
* Our dataset has reduced to 43875 rows in total. The 25th, 50th and 75th percentiles are close to their intial values before cleaning.
* The standard deviation(3837) is considerably lower than the value recorded before (480,000 USD).
* 16350 is a better approximation for the maximum price than the previous value of 100,000,000 USD, considering that our 75th percentile is around 6000 USD.

Let's now explore the date columns to understand the date range in data covers.

## Exploring the Date columns
There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted.

In [19]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

**Observations:**
* It appears that the first ten characters in each column represent date in yyyy-mm-dd format. The other characters represent time.

To understand the distribution we can extract just the date values by using a string method. We can use `Series.str[:10]` to select the first 10 characters  in each column.

In [20]:
print(autos['date_crawled'].str[:10].unique().size) #counts the unique crawl dates

date_crawled = (autos['date_crawled']
                     .str[:10]
                     .value_counts(normalize=True, dropna=False)
                     .round(3)
                     .sort_index()
                )

# use tabulate to print a pretty table
print(tabulate(date_crawled.to_frame(),headers=['Date','% crawled'], tablefmt='psql'))

# visualize data

#find a way to visualize data

34
+------------+-------------+
| Date       |   % crawled |
|------------+-------------|
| 2016-03-05 |       0.026 |
| 2016-03-06 |       0.014 |
| 2016-03-07 |       0.036 |
| 2016-03-08 |       0.033 |
| 2016-03-09 |       0.033 |
| 2016-03-10 |       0.033 |
| 2016-03-11 |       0.033 |
| 2016-03-12 |       0.037 |
| 2016-03-13 |       0.016 |
| 2016-03-14 |       0.037 |
| 2016-03-15 |       0.034 |
| 2016-03-16 |       0.03  |
| 2016-03-17 |       0.032 |
| 2016-03-18 |       0.013 |
| 2016-03-19 |       0.034 |
| 2016-03-20 |       0.038 |
| 2016-03-21 |       0.037 |
| 2016-03-22 |       0.032 |
| 2016-03-23 |       0.033 |
| 2016-03-24 |       0.029 |
| 2016-03-25 |       0.032 |
| 2016-03-26 |       0.033 |
| 2016-03-27 |       0.031 |
| 2016-03-28 |       0.035 |
| 2016-03-29 |       0.034 |
| 2016-03-30 |       0.034 |
| 2016-03-31 |       0.032 |
| 2016-04-01 |       0.033 |
| 2016-04-02 |       0.035 |
| 2016-04-03 |       0.039 |
| 2016-04-04 |       0.036 |
| 2016-04-0

In [26]:
date_crawled.describe()


count    34.000000
mean      0.029441
std       0.009792
min       0.001000
25%       0.030250
50%       0.033000
75%       0.034750
max       0.039000
Name: date_crawled, dtype: float64

**Observations**
* It appears the crawler collected information from the site everyday between March 5, 2016 and April 7, 2016, a period of 34 days.
* Some occassional drops can be observed on some weekend days in March (6th, 13th, 18th); and the last 3 days in April. However, the number of ads crawled seem evenly distributed overall.

In [21]:
ad_created = (autos['ad_created']
                     .str[:10]
                     .value_counts(normalize=True, dropna=False)
                     .round(3)
                     .sort_index()
                )
# Examine the earliest and last ad created dates.
print('Earliest Date', ad_created.head(1))
print('')
print('Last Date', ad_created.tail(1))



Earliest Date 2015-08-10    0.0
Name: ad_created, dtype: float64

Last Date 2016-04-07    0.001
Name: ad_created, dtype: float64
